### DISCUSSION TITLE
Using AJAX to call /oauth2/auth#

### DISCUSSION THREAD

**Discussion by RareSecond** [ory/hydra#2543] [Q&A] [answered]
> You cannot call /oauth2/auth using an AJAX request. It is not allowed and not possible with OAuth2. This endpoint can only be accessed using a normal browser request by clicking a link or redirecting the end-user's browser to that endpoint.

I know your docs state this, but why exactly is this? Because we are currently setting up Cypress, but we keep on getting the `No CSRF value available in the session cookie.` error.

However, Cypress also states that, especially when dealing with authentication, you **should** do it via `cy.request` (see https://docs.cypress.io/guides/getting-started/testing-your-app#Bypassing-your-UI).

I managed to get an `access_token` by using cy.request, but then I stumbled across your docs and it explicitly said not to use AJAX (or anything related to that I suppose). So, why not?

**Comment by aeneasr** [comment, marked as answer]
OAuth2 is a protocol which requires user interaction and browser redirects. Therefore, AJAX - which does not involve user interaction and does not redirect browsers - is the wrong methodology

For Cypress you will need to call the auth URL using cy.visit and probably persist the cookie across tests.

Also, afaik, cy.requests works a bit different from regular browser AJAX.

**Reply by RareSecond** [reply]
@aeneasr Two questions:
* So, what you guys mean in the docs, is that you **should not** call it with AJAX requests, because apparently it **is a possibility**.
* Could you explain then why I can't use Cypress to follow the login flow and instead get `No CSRF value available in the session cookie.`

**Reply by vinckr** [reply]
You should not call it with AJAX requests because it is not in accordance with the OAuth2.0 protocol. 
We word it like it is impossible to discourage people from cutting corners and implementing their own little workarounds. 
If you start doing that the security of the protocol and Hydra gets compromised and all sorts of terrible things can happen down the line.

Can you post some more information maybe on the Cypress flow?

How are you running hydra, in which environment? 
Can you post some configuration and maybe screenshots of the cookie header? 
Ideal would be a repo and steps to reproduce!

**Reply by aeneasr** [reply]
We are saying you can’t use it because the user needs to click a button „allow access to app foo“ in order to complete the flow.

**Reply by aeneasr** [reply]
You can also check our tests: https://github.com/ory/hydra/tree/master/cypress

**Reply by vinckr** [reply]
Hey @JDansercoer ,
Just checking in if your question was answered or if there was still something unclear.
I will mark this as answered for now, but feel free to come back to the discussion any time.

Thanks a ton! 🐝

**Reply by vsmbee** [reply]
@JDansercoer, could you please share your solution for _No CSRF value available in the session cookie._ ?
I complete understand and fully agree with the role of the user agent in OAuth2 and OIDC but am struggling to obtain the said `oauth2_authentication_session` cookie and implement a mock for E2E performance testing.
